src/abstractprobprog.jl

```diff
@@ -121,3 +122,28 @@ end
 function Base.rand(model::AbstractProbabilisticProgram)
     return rand(Random.default_rng(), NamedTuple, model)
 end
+
+"""
+    predict(
+        [rng::AbstractRNG=Random.default_rng(),]
+        [T=NamedTuple,]
+        model::AbstractProbabilisticProgram,
+        params,
+    ) -> T
+
+Draw a sample from the predictive distribution specified by `model` with its parameters fixed to `params`.
+
+The sample will be returned as format specified by `T`.
+"""
+function StatsBase.predict(rng::AbstractRNG, T::Type, model::AbstractProbabilisticProgram, params)
+    return rand(rng, T, fix(model, params))
+end
+function StatsBase.predict(T::Type, model::AbstractProbabilisticProgram, params)
```

Comment:
@torfjelde @yebai @penelopeysm do you think type `T` still a good idea?

I think it's probably okay for the function to be a bit under-speced now, so DynamicPPL and JuliaBUGS and others can decide what type to return.

Reply:
I think when Seth added this, it was modeled after the `rand` interface
https://github.com/TuringLang/AbstractPPL.jl/blob/b72a96368dc247549b9bbe10e13282ccd003517c/src/abstractprobprog.jl#L108-L124

I am for just having a simple `predict` now, or at most with `rng`, but not output type `T`.

Moreover, should we slim down the `rand` interface also, as this is going to be a breaking release.

Reply:
If we want DynamicPPL interface to have a type argument `T`, it makes sense for this one to have `T` as well. Otherwise I don't really see the point of having some interface here and then giving it a different signature in DynamicPPL.

Reply:
Agree with this, my comments only reflects that we don't have T argument right now if I recall correctly

Reply:
Ohhh, I see! For a general interface, though, if we don't specify `T` then we have to choose a privileged output type (like NamedTuple) right? Otherwise if it can return anything then it's not super useful either.

What do you think of something like this:

```julia
# default_return_type(model) specifies the default type returned by
# rand([rng, ]model) and predict([rng, ]model, params)
function default_return_type end

# Then we can have rand like this
function Base.rand(
    rng::Random.AbstractRNG = Random.default_rng(),
    ::Type{T} = default_return_type(model),
    model::AbstractProbabilisticProgram)
)
     # whatever implementation
end

# And predict like this
function StatsBase.predict(
    rng::Random.AbstractRNG = Random.default_rng(),
    ::Type{T} = default_return_type(model),
    model::AbstractProbabilisticProgram),
    params
)
     # whatever implementation
end 
```

Reply:
Basically, I don't feel super comfortable only being able to return NamedTuple. I think that's fine for a downstream package like DynamicPPL to choose, but enforcing it at this level might be too limiting. I also know I'm possibly overcomplicating things, sorry 😄

Reply:
@penelopeysm really sorry for missing the previous reply, your comments makes perfect sense to me. I think it's a good idea to let user decide what type to return. My view is: it might also not be a bad idea to let PPL packages that implements `AbstractPPL` decide what type to return, or return the same type as `params`.

I am just a bit uneasy to have a default return type that we don't always support (`OrderedDict` would be better, but also not perfect).